Stop using libstdc++ - win additional 44KB (on master branch)

risingPhil opened the pull request:
Hi,

Sorry, I couldn't really let go. ;-) Even though I saw that you already did some binary size optimizations, I got curious about the space consumption of the binary. So I investigated.

![image](https://github.com/user-attachments/assets/01ee3224-191d-4271-8b7e-0453b33dd858)


TL;DR: I managed to optimize the binary size and gain 44 KB by eliminating libstdc++.

Turns out the libraries used by the rom were consuming a significant amount of space (82 KB!) in your rom. (to see this, I added a compile option to generate a .map file. Then you can use the [amap](https://web.archive.org/web/20250104054010/https://www.sikorskiy.net/info/prj/amap/) tool to explore this .map file with a GUI)

![image](https://github.com/user-attachments/assets/13794f41-d110-4503-91c5-a255537073b8)

In fact, look at the sizes of the .rodata and .text sections here:

![image](https://github.com/user-attachments/assets/58b1ac70-363e-4d96-9510-74a3bf8f576c)
(.text = code, .rodata = readonly data)

Now, the thing is: if the linker adds things from libstdc++, it also adds the dependencies of those things from other libraries.

Anyway, I started optimizing libstdc++ away by taking out std::string, std::vector and std::to_string().

I did so by implementing my own versions of std::to_string() and std::vector -> ptgb::to_string() and ptgb::vector.

But it didn't end there. I tried completely eliminating it with the -nostdlib++. But turns out doing so also removes the startup routine from libgba. 

So I removed that compile option again. Turns out it wasn't really needed after all. Just removing any and all use of libstdc++ was enough to remove the library entirely.

I also had to define a custom operator new and operator delete to avoid pulling in things related to exception handling.

But this wasn't everything: after doing so, I noticed that libc still linked in a 13KB symbol called "categories" from libc_a-categories.o

![image](https://github.com/user-attachments/assets/f9b768ed-280c-4898-b0a1-359ad33087a0)

More investigation revealed that it was pulled in by the linker because libtonc was using siscanf() in tte_iohook.c. This usage of the function pulled in all the scanf related functions and locale support within libc.

I managed to eliminate it by defining a custom implementation of siscanf() that did JUST enough to keep tte_iohook.c working without pulling in any external dependencies.

I did have to add the nanoprintf library in order to use snprintf and vsnprintf without pulling in too many dependencies in libc. (more locale support and things like that).

The result:

As stated earlier, I managed to shave off 44 KB this way:

![image](https://github.com/user-attachments/assets/b72c3df9-8045-4aac-8db3-dcf1b545d758)

As you can see, libstdc++ is completely gone and libc is reduced to 9,5 KB!

In fact, all the libs together are only 35KB of the binary size now!

Consequences:

After this would get merged, I would advise staying away from libstdc++ to avoid pulling in bloat. That means stl datastructures, std::string and even adding support for initializer lists on custom classes. (it's fine for primitives though).

Don't worry about using basic C++ features though. Defining classes or doing inheritance is fine. And you can use libc functions to replace a lot of functionality included in libstdc++.

risingPhil committed:
Reduce binary size by eliminating libstdc++

This commit removes all references to things in the libstdc++ library to remove a decent chunk of bloat.

This means every std::to_string() call, std::string and std::vector. (as well as iostream related stuff).

I replaced those with my own versions ptgb::to_string() and ptgb::vector. Especially the latter is not exactly the same,
but close enough.

I also replaced operator new and delete with my own implementation to avoid pulling in everything related to exceptions
from libstdc++

Another problem was the fact that libtonc uses siscanf, which pulls in everything related to the scanf family of functions
and locale support. The worst part of that was that it included a 13KB "categories" symbol from libc_a-categories.o,
which was pulled in because of the locale support integrated into newlibc's siscanf() function. To fix that, I created a
custom, extremely restricted implementation of siscanf. libtonc only used this function to parse at most 2 integers from a
string anyway.

risingPhil committed:
Merge branch 'master' into optimizations/reduce-binarysize-remove-libstdc++

risingPhil commented:
Here are the .map files:

[output_after.map.zip](https://github.com/user-attachments/files/19673018/output_after.map.zip)
[output_before.map.zip](https://github.com/user-attachments/files/19673019/output_before.map.zip)

You can already see the difference after extraction of the .map files. The "before" map file is more than 5MB, the "after" map file is only 493KB!

Also: here is a zip with both the compiled multiboot rom and the associated loader.gba:

[Poke_Transporter_built_roms_optimized.zip](https://github.com/user-attachments/files/19673043/Poke_Transporter_built_roms_optimized.zip)

risingPhil commented:
@GearsProgress I do need to ask you to do some tests before merging this though.

Even though I doubt any regressions would get introduced by this, it would be best if you could do some tests.

The application boots fine with me, but I don't have cartridges with saves with significant enough progress to actually test the transfer process. Could you try this?

risingPhil committed:
Remove useless reinterpret_cast

GearsProgress commented:
(Obligatory apology for being late to respond to this- I usually only get the chance to work on this during the weekend, but I'm also sure you understand since you work on similar stuff haha)

This is... absolutely insane- I cannot thank you enough for all the effort you put in here!! ROM space is by far the biggest limiting factor with this project, so cutting it down by so much means this project can go even further.

I did change up most of the string implementation in a recent push, in part because of the way I implemented the different languages, so removing strings in their entirety shouldn't be *too* hard. Vectors on the other hand might be a bit tricky, so having the ptgb::vector is perfect. I'd love to work together to figure out how to squeeze as much space out of this as possible! I will fully admit that a lot of this code that we've been going through is pretty old and from when I was still learning how to optimize for the GBA, so there's some pretty questionable decisions I made haha.

The way I set up the movesets was to have 32 bytes for each Pokemon, and each bit corresponding to one of the 251 moves in gen 1/2. The bit would be true if the Pokemon can learn the move, and false if not. It would probably make more sense to instead have a list of all the Pokemon that can learn a certain move, that way we only need to store the data that is relevant. Like for instance, we can say "Mew and Ditto learn transform" instead of "here's 2 Pokemon that learn it and 249 that don't". *Really* what I should do though is just access the moveset data on the cartridge in real time (something I've been messing around with lately), and then I wouldn't need to store any of the data. Only issue is that that would be the gen 3 learnsets, not gen 2- but I'm not sure if they're even really that different. I'll have to look into that

GearsProgress commented:
Ah okay, after thinking about it for a grand total of five more minutes, I do check for the gen 3 movesets! I just don't have any moves introduced in gen 3 in the data because... they'll never show up lol. So I can use the gen 3 data within the ROMs. However, if there's any moveset differences between the gen 3 ROMs, it will depend on which one is currently inserted. Not sure if there is any difference though.

risingPhil commented:
Hi, don't worry about being late ;) I sure do understand :)

Yeah, your work on the string changes may have been one of the reasons why it was reasonably easy to remove the references to std::string. However, if it would've been necessary, I could just have written a ptgb::string as well. But overall, it's probably better to stick to c-strings instead.

I'd be happy to help you optimize more. Although I don't have much experience coding/optimizing for the GBA. So we'll see how far that would bring us :)

But I had a short in my brain about the MOVESETS. For some reason 32 bytes seemed to be a lot to me per pokémon. But I was thinking about the number of bits needed to represent the maximum value. But that number of bits obviously wouldn't be enough to store a bit representation of every move per pokémon. So now it finally clicked for me.

Reading the data straight from the rom might indeed be an option. And it's actually the approach I use for PokeMe64 (and libpokemegb). However, it would be annoying to figure out, because the rom offset for this table is likely different between versions (Ruby/Saphire vs Emerald vs Fire Red/Leaf Green), but also between the 6 languages. And keep in mind: the decompilation projects from Pret only target the English releases. They won't help you figure out the offsets of the same table in the various localizations. The way I handled that in libpokemegb (and therefore PokeMe64) is to write a [small program](https://github.com/risingPhil/libpokemegb/blob/main/examples/findLocalizationOffsets/main.cpp) that would look for a binary "pattern" and print the offset in the terminal. That way I could let it read some bytes from the english release at the "known offset" and look for the same bytes in the localized versions. But it's a hassle.

If you go for it, you might probably eliminate quite a few of these arrays in pokemon_data.cpp. But it could take time to go through each game variant. Especially for text tables (like names).

I'm also wondering if more "manual" optimizations can be done in regards to the sprites. I see various sprites seemingly being much larger than they need to be (in terms of the original PNG) (background.png, fennelBG.png, boxBG.png, openingBG.png). And with "larger than they need to be", I mean that a lot of these have parts that simply repeat a lot or are just straight up empty. (for instance, wouldn't the fennelGB be smaller if you just made it as big as fennel and store the "emotions" in a separate sprite?) But I have no idea how much this is compensated for by the grit tool and therefore whether it's worthwhile to fix that. These sprites do still seem rather large to me in the map file though. (but I have no experience with it, so it might be fine)

I might also be able to optimize away the verdana9 and sys8Glyphs fonts used as defaults by libtonc for another 4KB.

And if I disable some (currently unused) functionality of the nanoprintf functions, I might win another KB.

Anyway, I suppose it's irrelevant for this pull request right now. I think with the space we've already won (including @easyaspi314 's improvements), it should be a while before you run into this storage limitation again. :)

risingPhil commented:
You know what? You said you had been messing around with reading data from the rom.

Could you push those experiments to a separate branch? I want to have a look into this and those experiments could be a good starting point for me.

But regardless if I'd do changes like that, I'd open a separate pull request for them.

GearsProgress commented:
Ah! So fun thing about reading from the ROM- I've had that set up for a while, since it's needed in order to do what I'm doing in the custom Mystery Gift event! I have a file for each language that includes each version of the game, inside of `include/gba_rom_values`. This struct is then used to initialize the values of the 'rom_data' class in main, which is then used by a whole bunch of things throughout the code. This is definitely another thing that I could optimize and set up in a *much* cleaner way, but I haven't gotten to it yet. Currently I have the sprites being pulled in real time for Emerald, but I haven't gone in and added the values for the other games yet iirc.

As far as searching for the values, I have a python script that will compare the code between the known decomp locations and find it in the other languages. It isn't perfect, but it works well enough for me to have found all the locations I've needed thus far-- not super excited about having to find a lookup table though since that'll be completely different between games lol. At this point, there's probably no reason not to use all that data straight from the cartridge like you said- it would save a *ton* of space. The only thing I'm slightly iffy on is that moves are added and removed between RS/FRLG/E, but that also shouldn't be *that* big of a deal- and it honestly might not even be the case for any Gen 1/2 moves. I should write a little script that pulls all those moves and lists them off to me and see

GearsProgress commented:
After setting up my little python script, it seems like there are a decent number of moves that would differ between versions and wouldn't get picked up in certain games. It *probably* wouldn't be a big deal, but I'm not sure how I feel about it. There isn't really a good example of what to do here, since moves aren't ever removed, even in a VC transfer. But these moves really should be removed (at least in some cases) since they could prevent the Pokemon from being sent up in a future game. I'll have to think on it a bit. For now, here's the list of all the moves that were 'legal' in one game but not another:
[moves.txt](https://github.com/user-attachments/files/19737458/moves.txt)

risingPhil commented:
I think the solution is clear when you consider behavioral consistency between released versions of Poke Transfer GB:

Use the tables, but have special/override cases for these differences in your code. (Like: store a list of pairs (Pokémon id, allowed move id) for these cases in which it is only allowed in 1 game version.

Basically: ensure the behaviour remains the same as previous versions.

The list of differences should be small, I think.

risingPhil commented:
Nevermind, I just opened your list (I'm on my phone right now). It's much larger than I'd expect. 

Perhaps we should just consider some form of compression instead (such as RLE)

GearsProgress commented:
I just realized I didn't filter out the gen 3 moves, so there might be way less than that list. Let me run it again and see what we get.

If it's still super long, some kind of compression would definitely be the way to go, or some other way to store the data than a bit list

risingPhil commented:
Still, when I glanced over the data, I saw moves like snore in there. So I have a feeling this list is longer than we'd expect.

But there are bound to be a lot of zeros in your current MOVESETS array. So... Something like RLE, similar to how it was used for gen 1 sprites should be a decent solution.

You could also consider store a list of allowed move ids per Pokémon instead.

But the question is how beneficial that approach would be considering every moveid would occupy a whole byte.

But I suppose we could just test all these approaches.

GearsProgress commented:
Just ran the script and it was exactly the same, you called it lol

That's what I was worried about with the list too- it would be a byte for each move rather than a bit. It would be smaller for Pokémon that learn less than 32 moves, but longer otherwise. RLE compression would probably be the best way forward then. I'm currently messing around with the transfer algorithm, but once that's all fixed up I'd love to set up that compression. Honestly, any Pokémon data that can't be pulled from the cartridge should probably be compressed like that to save space.

risingPhil commented:
So... I did a RLE implementation similar to the one used for gen I sprites. However, the result is disappointing for the MOVESETS table at least:

`philippe@zapptop106:~/Downloads/testRLE$ ./testRLE 
Pokemon 0: RLE -> 2 bytes
                PASS!
Pokemon 1: RLE -> 26 bytes
                PASS!
Pokemon 2: RLE -> 25 bytes
                PASS!
Pokemon 3: RLE -> 27 bytes
                PASS!
Pokemon 4: RLE -> 29 bytes
                PASS!
Pokemon 5: RLE -> 29 bytes
                PASS!
Pokemon 6: RLE -> 30 bytes
                PASS!
Pokemon 7: RLE -> 30 bytes
                PASS!
Pokemon 8: RLE -> 29 bytes
                PASS!
Pokemon 9: RLE -> 30 bytes
                PASS!
Pokemon 10: RLE -> 5 bytes
                PASS!
Pokemon 11: RLE -> 4 bytes
                PASS!
Pokemon 12: RLE -> 24 bytes
                PASS!
Pokemon 13: RLE -> 5 bytes
                PASS!
Pokemon 14: RLE -> 4 bytes
                PASS!
Pokemon 15: RLE -> 25 bytes
                PASS!
Pokemon 16: RLE -> 24 bytes
                PASS!
Pokemon 17: RLE -> 22 bytes
                PASS!
Pokemon 18: RLE -> 23 bytes
                PASS!
Pokemon 19: RLE -> 28 bytes
                PASS!
Pokemon 20: RLE -> 28 bytes
                PASS!
Pokemon 21: RLE -> 25 bytes
                PASS!
Pokemon 22: RLE -> 24 bytes
                PASS!
Pokemon 23: RLE -> 25 bytes
                PASS!
Pokemon 24: RLE -> 24 bytes
                PASS!
Pokemon 25: RLE -> 28 bytes
                PASS!
Pokemon 26: RLE -> 28 bytes
                PASS!
Pokemon 27: RLE -> 26 bytes
                PASS!
Pokemon 28: RLE -> 26 bytes
                PASS!
Pokemon 29: RLE -> 27 bytes
                PASS!
Pokemon 30: RLE -> 26 bytes
                PASS!
Pokemon 31: RLE -> 30 bytes
                PASS!
Pokemon 32: RLE -> 28 bytes
                PASS!
Pokemon 33: RLE -> 26 bytes
                PASS!
Pokemon 34: RLE -> 30 bytes
                PASS!
Pokemon 35: RLE -> 32 bytes
                PASS!
Pokemon 36: RLE -> 32 bytes
                PASS!
Pokemon 37: RLE -> 24 bytes
                PASS!
Pokemon 38: RLE -> 23 bytes
                PASS!
Pokemon 39: RLE -> 30 bytes
                PASS!
Pokemon 40: RLE -> 30 bytes
                PASS!
Pokemon 41: RLE -> 25 bytes
                PASS!
Pokemon 42: RLE -> 24 bytes
                PASS!
Pokemon 43: RLE -> 22 bytes
                PASS!
Pokemon 44: RLE -> 22 bytes
                PASS!
Pokemon 45: RLE -> 22 bytes
                PASS!
Pokemon 46: RLE -> 27 bytes
                PASS!
Pokemon 47: RLE -> 25 bytes
                PASS!
Pokemon 48: RLE -> 25 bytes
                PASS!
Pokemon 49: RLE -> 25 bytes
                PASS!
Pokemon 50: RLE -> 23 bytes
                PASS!
Pokemon 51: RLE -> 23 bytes
                PASS!
Pokemon 52: RLE -> 28 bytes
                PASS!
Pokemon 53: RLE -> 29 bytes
                PASS!
Pokemon 54: RLE -> 29 bytes
                PASS!
Pokemon 55: RLE -> 29 bytes
                PASS!
Pokemon 56: RLE -> 29 bytes
                PASS!
Pokemon 57: RLE -> 29 bytes
                PASS!
Pokemon 58: RLE -> 24 bytes
                PASS!
Pokemon 59: RLE -> 24 bytes
                PASS!
Pokemon 60: RLE -> 25 bytes
                PASS!
Pokemon 61: RLE -> 29 bytes
                PASS!
Pokemon 62: RLE -> 28 bytes
                PASS!
Pokemon 63: RLE -> 27 bytes
                PASS!
Pokemon 64: RLE -> 29 bytes
                PASS!
Pokemon 65: RLE -> 29 bytes
                PASS!
Pokemon 66: RLE -> 30 bytes
                PASS!
Pokemon 67: RLE -> 28 bytes
                PASS!
Pokemon 68: RLE -> 29 bytes
                PASS!
Pokemon 69: RLE -> 26 bytes
                PASS!
Pokemon 70: RLE -> 23 bytes
                PASS!
Pokemon 71: RLE -> 23 bytes
                PASS!
Pokemon 72: RLE -> 26 bytes
                PASS!
Pokemon 73: RLE -> 24 bytes
                PASS!
Pokemon 74: RLE -> 25 bytes
                PASS!
Pokemon 75: RLE -> 25 bytes
                PASS!
Pokemon 76: RLE -> 28 bytes
                PASS!
Pokemon 77: RLE -> 23 bytes
                PASS!
Pokemon 78: RLE -> 25 bytes
                PASS!
Pokemon 79: RLE -> 29 bytes
                PASS!
Pokemon 80: RLE -> 33 bytes
                PASS!
Pokemon 81: RLE -> 22 bytes
                PASS!
Pokemon 82: RLE -> 23 bytes
                PASS!
Pokemon 83: RLE -> 26 bytes
                PASS!
Pokemon 84: RLE -> 26 bytes
                PASS!
Pokemon 85: RLE -> 24 bytes
                PASS!
Pokemon 86: RLE -> 25 bytes
                PASS!
Pokemon 87: RLE -> 21 bytes
                PASS!
Pokemon 88: RLE -> 26 bytes
                PASS!
Pokemon 89: RLE -> 27 bytes
                PASS!
Pokemon 90: RLE -> 24 bytes
                PASS!
Pokemon 91: RLE -> 22 bytes
                PASS!
Pokemon 92: RLE -> 23 bytes
                PASS!
Pokemon 93: RLE -> 21 bytes
                PASS!
Pokemon 94: RLE -> 29 bytes
                PASS!
Pokemon 95: RLE -> 25 bytes
                PASS!
Pokemon 96: RLE -> 28 bytes
                PASS!
Pokemon 97: RLE -> 28 bytes
                PASS!
Pokemon 98: RLE -> 28 bytes
                PASS!
Pokemon 99: RLE -> 28 bytes
                PASS!
Pokemon 100: RLE -> 22 bytes
                PASS!
Pokemon 101: RLE -> 23 bytes
                PASS!
Pokemon 102: RLE -> 24 bytes
                PASS!
Pokemon 103: RLE -> 25 bytes
                PASS!
Pokemon 104: RLE -> 31 bytes
                PASS!
Pokemon 105: RLE -> 30 bytes
                PASS!
Pokemon 106: RLE -> 27 bytes
                PASS!
Pokemon 107: RLE -> 26 bytes
                PASS!
Pokemon 108: RLE -> 32 bytes
                PASS!
Pokemon 109: RLE -> 26 bytes
                PASS!
Pokemon 110: RLE -> 26 bytes
                PASS!
Pokemon 111: RLE -> 28 bytes
                PASS!
Pokemon 112: RLE -> 29 bytes
                PASS!
Pokemon 113: RLE -> 33 bytes
WARN: can't write 00
WARN: can't write 00
                PASS!
Pokemon 114: RLE -> 26 bytes
                PASS!
Pokemon 115: RLE -> 34 bytes
                FAILED!
Pokemon 116: RLE -> 26 bytes
                PASS!
Pokemon 117: RLE -> 22 bytes
                PASS!
Pokemon 118: RLE -> 23 bytes
                PASS!
Pokemon 119: RLE -> 23 bytes
                PASS!
Pokemon 120: RLE -> 24 bytes
                PASS!
Pokemon 121: RLE -> 25 bytes
                PASS!
Pokemon 122: RLE -> 29 bytes
                PASS!
Pokemon 123: RLE -> 26 bytes
                PASS!
Pokemon 124: RLE -> 30 bytes
                PASS!
Pokemon 125: RLE -> 29 bytes
                PASS!
Pokemon 126: RLE -> 27 bytes
                PASS!
Pokemon 127: RLE -> 25 bytes
                PASS!
Pokemon 128: RLE -> 26 bytes
                PASS!
Pokemon 129: RLE -> 6 bytes
                PASS!
Pokemon 130: RLE -> 24 bytes
                PASS!
Pokemon 131: RLE -> 27 bytes
                PASS!
Pokemon 132: RLE -> 4 bytes
                PASS!
Pokemon 133: RLE -> 22 bytes
                PASS!
Pokemon 134: RLE -> 26 bytes
                PASS!
Pokemon 135: RLE -> 26 bytes
                PASS!
Pokemon 136: RLE -> 25 bytes
                PASS!
Pokemon 137: RLE -> 27 bytes
                PASS!
Pokemon 138: RLE -> 26 bytes
                PASS!
Pokemon 139: RLE -> 25 bytes
                PASS!
Pokemon 140: RLE -> 26 bytes
                PASS!
Pokemon 141: RLE -> 26 bytes
                PASS!
Pokemon 142: RLE -> 28 bytes
                PASS!
Pokemon 143: RLE -> 31 bytes
                PASS!
Pokemon 144: RLE -> 25 bytes
                PASS!
Pokemon 145: RLE -> 26 bytes
                PASS!
Pokemon 146: RLE -> 25 bytes
                PASS!
Pokemon 147: RLE -> 26 bytes
                PASS!
Pokemon 148: RLE -> 24 bytes
                PASS!
Pokemon 149: RLE -> 30 bytes
                PASS!
Pokemon 150: RLE -> 33 bytes
WARN: can't write 00
WARN: can't write 00
WARN: can't write 00
WARN: can't write 00
                PASS!
Pokemon 151: RLE -> 36 bytes
WARN: can't write 00
                FAILED!
Pokemon 152: RLE -> 26 bytes
                PASS!
Pokemon 153: RLE -> 25 bytes
                PASS!
Pokemon 154: RLE -> 26 bytes
                PASS!
Pokemon 155: RLE -> 24 bytes
                PASS!
Pokemon 156: RLE -> 25 bytes
                PASS!
Pokemon 157: RLE -> 28 bytes
                PASS!
Pokemon 158: RLE -> 28 bytes
                PASS!
Pokemon 159: RLE -> 29 bytes
                PASS!
Pokemon 160: RLE -> 29 bytes
                PASS!
Pokemon 161: RLE -> 30 bytes
                PASS!
Pokemon 162: RLE -> 30 bytes
                PASS!
Pokemon 163: RLE -> 28 bytes
                PASS!
Pokemon 164: RLE -> 26 bytes
                PASS!
Pokemon 165: RLE -> 27 bytes
                PASS!
Pokemon 166: RLE -> 26 bytes
                PASS!
Pokemon 167: RLE -> 26 bytes
                PASS!
Pokemon 168: RLE -> 24 bytes
                PASS!
Pokemon 169: RLE -> 24 bytes
                PASS!
Pokemon 170: RLE -> 23 bytes
                PASS!
Pokemon 171: RLE -> 24 bytes
                PASS!
Pokemon 172: RLE -> 26 bytes
                PASS!
Pokemon 173: RLE -> 32 bytes
                PASS!
Pokemon 174: RLE -> 29 bytes
                PASS!
Pokemon 175: RLE -> 31 bytes
                PASS!
Pokemon 176: RLE -> 31 bytes
                PASS!
Pokemon 177: RLE -> 26 bytes
                PASS!
Pokemon 178: RLE -> 26 bytes
                PASS!
Pokemon 179: RLE -> 22 bytes
                PASS!
Pokemon 180: RLE -> 27 bytes
                PASS!
Pokemon 181: RLE -> 28 bytes
                PASS!
Pokemon 182: RLE -> 22 bytes
                PASS!
Pokemon 183: RLE -> 29 bytes
                PASS!
Pokemon 184: RLE -> 27 bytes
                PASS!
Pokemon 185: RLE -> 27 bytes
                PASS!
Pokemon 186: RLE -> 28 bytes
                PASS!
Pokemon 187: RLE -> 25 bytes
                PASS!
Pokemon 188: RLE -> 22 bytes
                PASS!
Pokemon 189: RLE -> 23 bytes
                PASS!
Pokemon 190: RLE -> 31 bytes
                PASS!
Pokemon 191: RLE -> 22 bytes
                PASS!
Pokemon 192: RLE -> 23 bytes
                PASS!
Pokemon 193: RLE -> 26 bytes
                PASS!
Pokemon 194: RLE -> 27 bytes
                PASS!
Pokemon 195: RLE -> 29 bytes
                PASS!
Pokemon 196: RLE -> 26 bytes
                PASS!
Pokemon 197: RLE -> 26 bytes
                PASS!
Pokemon 198: RLE -> 27 bytes
                PASS!
Pokemon 199: RLE -> 31 bytes
                PASS!
Pokemon 200: RLE -> 25 bytes
                PASS!
Pokemon 201: RLE -> 3 bytes
                PASS!
Pokemon 202: RLE -> 7 bytes
                PASS!
Pokemon 203: RLE -> 29 bytes
                PASS!
Pokemon 204: RLE -> 26 bytes
                PASS!
Pokemon 205: RLE -> 25 bytes
                PASS!
Pokemon 206: RLE -> 28 bytes
                PASS!
Pokemon 207: RLE -> 26 bytes
                PASS!
Pokemon 208: RLE -> 26 bytes
                PASS!
Pokemon 209: RLE -> 29 bytes
                PASS!
Pokemon 210: RLE -> 30 bytes
                PASS!
Pokemon 211: RLE -> 26 bytes
                PASS!
Pokemon 212: RLE -> 25 bytes
                PASS!
Pokemon 213: RLE -> 24 bytes
                PASS!
Pokemon 214: RLE -> 25 bytes
                PASS!
Pokemon 215: RLE -> 31 bytes
                PASS!
Pokemon 216: RLE -> 29 bytes
                PASS!
Pokemon 217: RLE -> 29 bytes
                PASS!
Pokemon 218: RLE -> 24 bytes
                PASS!
Pokemon 219: RLE -> 25 bytes
                PASS!
Pokemon 220: RLE -> 25 bytes
                PASS!
Pokemon 221: RLE -> 26 bytes
                PASS!
Pokemon 222: RLE -> 26 bytes
                PASS!
Pokemon 223: RLE -> 24 bytes
                PASS!
Pokemon 224: RLE -> 24 bytes
                PASS!
Pokemon 225: RLE -> 26 bytes
                PASS!
Pokemon 226: RLE -> 26 bytes
                PASS!
Pokemon 227: RLE -> 27 bytes
                PASS!
Pokemon 228: RLE -> 28 bytes
                PASS!
Pokemon 229: RLE -> 27 bytes
                PASS!
Pokemon 230: RLE -> 23 bytes
                PASS!
Pokemon 231: RLE -> 23 bytes
                PASS!
Pokemon 232: RLE -> 24 bytes
                PASS!
Pokemon 233: RLE -> 28 bytes
                PASS!
Pokemon 234: RLE -> 29 bytes
                PASS!
Pokemon 235: RLE -> 4 bytes
                PASS!
Pokemon 236: RLE -> 24 bytes
                PASS!
Pokemon 237: RLE -> 25 bytes
                PASS!
Pokemon 238: RLE -> 30 bytes
                PASS!
Pokemon 239: RLE -> 27 bytes
                PASS!
Pokemon 240: RLE -> 27 bytes
                PASS!
Pokemon 241: RLE -> 30 bytes
                PASS!
Pokemon 242: RLE -> 33 bytes
                PASS!
Pokemon 243: RLE -> 27 bytes
                PASS!
Pokemon 244: RLE -> 29 bytes
                PASS!
Pokemon 245: RLE -> 26 bytes
                PASS!
Pokemon 246: RLE -> 23 bytes
                PASS!
Pokemon 247: RLE -> 20 bytes
                PASS!
Pokemon 248: RLE -> 28 bytes
                PASS!
Pokemon 249: RLE -> 30 bytes
                PASS!
Pokemon 250: RLE -> 29 bytes
                PASS!
Pokemon 251: RLE -> 27 bytes
                PASS!
Pokemon 252: RLE -> 4 bytes
                PASS!
Total: 6466 bytes`

Here is the source code: [testRLE.zip](https://github.com/user-attachments/files/19763776/testRLE.zip)

Turns out there are more non-zero bits/bytes than I expected.

So it looks like this approach is not viable.

However, in a conversation with chatGPT, the ZX0 compression algorithm came up. And it turns out the implementation would be tiny and it should have a decent compression ratio. (according to ChatGPT we should be able to compress this MOVESETS table to between ~4100 to 4300 bytes).

So I'm going to look into that.

risingPhil committed:
Merge branch 'master' into optimizations/reduce-binarysize-remove-libstdc++

risingPhil commented:
So, small update.

I have implemented zx0 on this branch: https://github.com/risingPhil/Poke_Transporter_GB/tree/optimizations/implement-zx0-compression

It basically compresses all the tables from pokemon_data.cpp. This data went from 18 KB to 12 KB. (not all of it compresses well, and some can be optimized further). The MOVESETS table in particular went from 8 KB to 5 KB.

But there are many more compression opportunities (which I haven't integrated yet):

- the rom_data structs
- the GB_ROM structs
- translated_text.cpp

By the way, while I haven't implemented it yet, I have an idea on how to optimize the MOVESETS table further: We can have the evolved pokémon only store the bits that are different in comparison to the base form. And not only that, we can deduplicate the unique entries and have a separate table that stores a rowindex for each pokémon.

Anyway, there is still a problem: the ZX0 decompression performance is not yet where it needs to be. Decoding a 512 byte array (one of those charsets) currently leads to a visible delay. (the "debug" credits screen). But I know the mystery_gift_builder related stuff will be much worse. (I can't test it at the moment)

I know I can still optimize the number of bytes I read at a time, so it's not like I'm at a dead end at the moment.

But I am still worried about the mystery_gift_builder code in terms of performance. (because it needs SO many of these tables, so I'm worried it would freeze for too long during this process)

Right now, I'm just going to try to get performance to the right level.

Now, as an alternative, I might consider switching out ZX0 with the default LZ77 decompression if I don't get ZX0 to the desired performance. It might be faster that way and probably easier too.

Anyway, I still have a lot of work cut out for me before I can create a pull request for it.

risingPhil commented:
I just wanted to share some stats about the rom occupations, because I thought you might find it interesting:

Keep in mind that these stats are from my (separate) [zx0 branch](https://github.com/risingPhil/Poke_Transporter_GB/tree/optimizations/implement-zx0-compression) though

libs: 34 KB
code: 36,5 KB

sprites: 23,4 KB
compressed data tables (pokemon_data): 11 KB
translated_text: 13,8 KB
unneeded fonts (sys8Glyphs+verdana9Glyphs): 4KB
latin_shortglyphs: 4 KB
rom_data structs: 6,9 KB
GB_ROM structs: 1,6 KB

various (uncounted): 12 KB

So for compression, like I said earlier, I think the rom_data, GB_ROM and translated_text are still opportunities for compression.

There are also still opportunities to optimize the compressed tables by changing the data representation (MOVESETS, TYPES) or eliminating some data entirely if possible. (for instance the unneeded fonts).

But I first need to get performance at the level it needs to be.

All of this is happening on the separate branch though. It doesn't affect this pull request at all. (But this pull request is currently my communication medium with you, so...)

GearsProgress commented:
This is fascinating, thank you! I also really like your idea about only adding the changes for an evolution, rather than all the moves again- that would be great!

I do have a Discord that I'd be happy to chat on (since yeah, this probably isn't the best place to discuss all this haha). My username is @starport75- let me know when you've sent over a request :)

risingPhil commented:
I just added you on discord. ;) Thank you for the username.

But in the meantime I have another update.

So I managed to optimize ZX0 compression significantly. I can now decompress 512 bytes in 1.3 ms. So that's around the performance level I'm happy with. But it's not all good. More on that later.

In fact, I also did the optimization of the MOVESETS table. Storing only the "changed" bits in relation to the base forms made it slightly better compressable. I won around 300 bytes. But unfortunately eliminating duplicate rows doesn't work because there aren't any duplicates, even with this change. I validated it by writing a small unit test. Still, I managed to compress the MOVESETS table down to 4641 bytes.

But I just realized today that there's a big problem with using compression for these data tables: We now need a lot more IWRAM!

The thing is: all global buffers and the stack are being stored in IWRAM.

And the thing is: your convert_to_gen_three() function uses about all of these data tables. When uncompressed and only keeping around the charsets we actually need, we would still require 11 KB of IWRAM (even if we allocate it on the stack).

So let's take a look at our IWRAM usage:

- .bss (uninitialized global vars) section is 12,3 KB
- .data (initialized global vars) section is 6,9 KB
- .iwram (special functions, including my new zx0 decompressor (which takes up 1,5 KB of this)) section is 3 KB

So in total, we were already using 22,2 KB of IWRAM (out of 32 KB). That only leaves 9,8 KB for the stack. So this 11 KB is out of the question. And the convert_to_gen_three() function is getting used in functions which already allocate a LOT of data on the stack. 

Now, you could think to just unload tables and decompress them again all the time. But the problem with that is that the convert_to_gen_three() function is getting called in loops that handle entire pokémon boxes. And given that it uses about every data table, it would have a big performance impact in your critical path.

And the MOVESETS table is also being used within that same flow. So we kinda need to keep it around when looping.

But... if the table is stored uncompressed like before my changes, it would've been part of EWRAM. So then we wouldn't run into that problem.

So actually, I think we should move the MOVESETS table back into the uncompressed form it was before to avoid filling up the remaining space in IWRAM at the point we actually really need it.

But that diminishes the effectiveness of the compression quite a lot, obviously, because the MOVESETS table was my main motivator to work on compression in the first place. 

But not all is lost. Like I said earlier, there are still more opportunities to compress (like the translated_text and rom data structs). And in those cases, we won't run into the IWRAM shortage.

And there's still more good news: I managed to optimize away a 4 KB "handles" buffer from libsysbase_libsysbase_a-handle_manager.o that was part of both the rom data AND IWRAM. (everything in the .data section is getting copied from the multiboot rom data to IWRAM at startup)

But yeah, to get some mileage out of the compression, I'll have to do some more work. ;)

risingPhil commented:
I changed the target branch to release-candidate as discussed in private.

risingPhil committed:
Merge remote-tracking branch 'upstream/text-and-box-rewrite' into optimizations/reduce-binarysize-remove-libstdc++

risingPhil committed:
Merge remote-tracking branch 'upstream/text-and-box-rewrite' into optimizations/reduce-binarysize-remove-libstdc++

risingPhil committed:
Merge remote-tracking branch 'upstream/text-and-box-rewrite' into optimizations/reduce-binarysize-remove-libstdc++

risingPhil committed:
Fix regression

ptgb::vector::resize() had a bug which caused it to incorrectly add/erase elements.

The reason was a loop end condition that was modified during iterations of said loop. This end condition
was intended to be static, but because it used a variable that was modified during the loop, it didn't
repeat the intended number of times.

Fix: before starting the for loop, calculate the end condition once.

risingPhil committed:
Merge remote-tracking branch 'upstream/text-and-box-rewrite' into optimizations/reduce-binarysize-remove-libstdc++
